### ISSUE TITLE
More specific return is not allowed with generic return type

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#1878]
https://psalm.dev/r/2b59056447
```php
<?php
  /** @template T */
  class C {
    /** @param T $_val */
    public function __construct($_val) {}
  } 
  /** @return C<array<int,int>> */
  function f() {
    return new C([1]);
  }

  /** @return array<int,int> */
  function f2() {
    return [1]; // this is fine though
  }
```
**Expected**: no errors, as `C<array{0:int(1)}>` is a valid subtype of `C<array<int,int>>`, just like `array{0:int(1)}` is a subtype of `array<int,int>`
**Actual**:
```
Psalm output (using commit 791b30a): 

ERROR: InvalidReturnStatement - 9:12 - The type 'C<array{0: int(1)}>' does not match the declared return type 'C<array<int, int>>' for f

ERROR: InvalidReturnType - 7:15 - The declared return type 'C<array<int, int>>' for f is incorrect, got 'C<array{0: int}>'
```

Using `@return C<SomeInterface>` with `return new C<new class implements SomeInterface {}>` causes the same error, so it's not limited to just array type params.

**Comment by weirdan** [comment]
Aren't templates covariant by default anymore? Switching to `@template-covariant` fixes these issues :thinking:

**Comment by iluuu1994** [comment]
> Aren't templates covariant by default anymore?

No, that was changed recently. Psalm didn't distinguish between invariant and covariant types before. Template types are now invariant by default and can be marked as covariant which means you won't be able to use them as parameter types any more as that wouldn't be type safe.

**Comment by weirdan** [comment]
I suppose templated stubs shipped with Psalm (like iterators, ArrayObject and so on) were not reviewed when this change was introduced, right?

**Comment by iluuu1994** [comment]
They have. The element type of `ArrayObject` is invariant since the collection is mutable.

**Comment by weirdan** [comment]
And rationale for not allowing `ArrayObject<array-key, array{0:int}>` to be returned where return type is declared as `ArrayObject<array-key,array<int, int>>` would be that someone could unset that 0 offset from the outside while object holding the reference to it would still assume it's available?

**Comment by iluuu1994** [comment]
@muglug Just found this bug:

```php
<?php

class Animal {}
class Cat extends Animal {}

/**
 * @param \ArrayObject<int, Animal> $collection
 */
function takesAnimalCollection($collection): void {}

/**
 * @param iterator<int, Animal> $iterator
 */
function takesAnimalIterator($iterator): void {}

/** @var \ArrayObject<int, Cat> $catCollection */
$catCollection = new \ArrayObject();

takesAnimalCollection($catCollection); // Argument 1 of takesAnimalCollection expects ArrayObject<int, Animal>, ArrayObject<int, Cat> provided
takesAnimalIterator($catCollection); // Argument 1 of takesAnimalCollection expects ArrayObject<int, Animal>, ArrayObject<int, Cat> provided
```

https://psalm.dev/r/8610463268

The last line shouldn't report an error. When you upcast to an interface where the template type is covariant the template type should also be allowed to be upcasted. When you change type of the original collection Psalm gets that it's type safe to upcast the template type.

https://psalm.dev/r/47e3beb715

**Comment by iluuu1994** [comment]
> And rationale for not allowing ArrayObject<array-key, array{0:int}> to be returned where return type is declared as ArrayObject<array-key,array<int, int>> would be that someone could unset that 0 offset from the outside while object holding the reference to it would still assume it's available, correct?

Yeah. You could add any array containing integer keys and values to the `ArrayObject` but then it wouldn't be of type `ArrayObject<array-key, array{0:int}>` any more. You should be allowed to pass it as `iterator<array-key, array{0:int}>` though because that would make it immutable in the scope you pass it to.

**Comment by muglug** [comment]
> Just found this bug:

@iluuu1994 not a bug - it doesn't work! https://3v4l.org/2nhMI - maybe you meant `iterable`?

@weirdan you can add an explicit `@var` above the return to make Psalm happy: https://psalm.dev/r/64817f8b29

**Comment by iluuu1994** [comment]
@muglug Of course, sorry. Works just fine.
